### PR TITLE
e2e: test labels

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Adding tags with the new framework helper functions also adds them as Ginkgo labels. Labels enable much more flexible [filtering of specs](https://onsi.github.io/ginkgo/#filtering-specs). The goal is to register all tests this way, with no or only minimal changes to the full test names, so existing regular expressions should continue to work.

There's one exception: a space now gets inserted around tags, which wasn't always the case before. If a regular expression in a Prow job is sensitive to white space, then it will not match anymore.

#### Which issue(s) this PR fixes:
Related-to:  https://github.com/kubernetes/enhancements/issues/3041

The KEP discusses semantic changes. Those are not part of this PR. Instead, the goal is to remain as close as possible to the current test names.

#### Special notes for your reviewer:

A subset was merged in https://github.com/kubernetes/kubernetes/pull/121391. Announced in https://groups.google.com/a/kubernetes.io/g/dev/c/DTFEng143NY

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by pohly** [comment]
/hold

To prevent accidentally merging it on the first lgtm.

**Comment by pohly** [comment]
The PR only adds spaces:
```console
# check out master
$ (go test -v ./test/e2e ./test/e2e_node ./test/e2e_kubeadm -args --list-tests ) >before
# check out PR
$ (go test -v ./test/e2e ./test/e2e_node ./test/e2e_kubeadm -args --list-tests ) >after
# compare, ignoring the source code lines (those change) and whitespaces
diff -w <(sed -e 's/ *\(\.*[/a-z0-9_-]*.go\):[0-9]*: /\1: /' before | sort) <(sed -e 's/ *\(\.*[/a-z0-9_-]*.go\):[0-9]*: /\1: /'  after | sort)
1640,1642c1640,1642
< ok  	k8s.io/kubernetes/test/e2e	0.613s
< ok  	k8s.io/kubernetes/test/e2e_kubeadm	0.067s
< ok  	k8s.io/kubernetes/test/e2e_node	0.123s
---
> ok  	k8s.io/kubernetes/test/e2e	0.621s
> ok  	k8s.io/kubernetes/test/e2e_kubeadm	0.060s
> ok  	k8s.io/kubernetes/test/e2e_node	0.125s
```

File: test/e2e/framework/ginkgowrapper.go [hunk 1]
```diff
@@ -246,6 +246,84 @@ func registerInSuite(ginkgoCall func(string, ...interface{}) bool, args []interf
 	return ginkgoCall(text, ginkgoArgs...)
 }
 
+var (
+	tagRe                 = regexp.MustCompile(`\[.*?\]`)
```
Comment (aojea) [inline test/e2e/framework/ginkgowrapper.go:250]: why do you need a question mark in this refect? the `\[.*\]` is not enough?

Comment (pohly) [reply test/e2e/framework/ginkgowrapper.go:250]: Without the question mark, two tags next to each other (`[foo][bar]`) get matched as one. The question mark ensures that matching stop as soon as the first closing `]` is found.

Comment (pohly) [reply test/e2e/framework/ginkgowrapper.go:250]: @aojea: okay to lift the hold?

Comment (aojea) [reply test/e2e/framework/ginkgowrapper.go:250]: I hate regex 😄

**Comment by aojea** [comment]
/milestone v1.29

Discussed in https://kubernetes.slack.com/archives/C09QZ4DQB/p1698155952275539

This has been approved and lgtmed from SIG Testing and other stakeholders, however, since this PR has a big impact on any feature developed during this cycle and most probably cause conflicts, instead of merge it soon and cause additional works to other features, we are going to unhold it on the last day so any possible rebase conflict will be absorved by this PR during the post-code freeze and "test fixing" period.

File: test/e2e/framework/ginkgowrapper.go (hunk 1 above)
Comment (aojea) [reply test/e2e/framework/ginkgowrapper.go:250]: > @aojea: okay to lift the hold?

yeah. added a comment, it seems my browser. didn't show me this comment before Xd

**Comment by MaryamTavakkoli** [comment]
Hello!
Bug triage for the 1.29 release cycle is here!
The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)). as we want to ensure that each PR has a chance to be merged.

Are we still targeting the 1.29 release?

**Comment by pohly** [comment]
> we are going to unhold it on the last day

So let's do that...

/hold cancel

**Comment by pohly** [comment]
Conflict in `test/e2e_node/pod_host_ips.go`, so this isn't going to get merged today. Let's keep the milestone and I'll fix this after code freeze.

**Comment by pohly** [comment]
@aojea : rebased, please lgtm again.

**Comment by aojea** [comment]
it seems I broke you `  ERROR: test/e2e/network/service_cidrs.go:38: [Feature:ServiceCIDRs] in plain text is deprecated and must be added through WithFeature(ServiceCIDRs) instead` 😄 

sorry about that

**Comment by pohly** [comment]
Hey, it's working as intended - good check! :grin: 

Let me fix that... would have been trivial if I had remembered to run `hack/verify-e2e-suites.sh`:
```
ERROR: E2E test suite invocation failed for test/e2e.
   ERROR: E2E suite initialization was faulty, these errors must be fixed:
   ERROR: test/e2e/network/service_cidrs.go:38: [Feature:ServiceCIDRs] in plain text is deprecated and must be added through WithFeature(ServiceCIDRs) instead
E2E suite test/e2e_kubeadm passed.
ERROR: E2E test suite invocation failed for test/e2e_node.
   ERROR: E2E suite initialization was faulty, these errors must be fixed:
   ERROR: test/e2e_node/container_lifecycle_test.go:2964: [Serial] in plain text is deprecated and must be added through WithSerial instead
```

Fixed: https://github.com/kubernetes/kubernetes/compare/2710abf664524c3f544bf5c2410669075febd72b..338fe7ad55cad858284d4c96df86d3cee9c7d29c

**Pull request merged**